outbox-kafka-spring/src/main/java/one/tomorrow/transactionaloutbox/service/OutboxProcessor.java

```diff
@@ -165,15 +163,31 @@ private void processOutboxWithLock() {
 
 	}
 
-	private void processOutbox() throws ExecutionException, InterruptedException {
-		List<OutboxRecord> records = repository.getUnprocessedRecords(BATCH_SIZE);
-		for (OutboxRecord outboxRecord : records) {
-			ProducerRecord<String, byte[]> producerRecord = toProducerRecord(outboxRecord);
-			Future<RecordMetadata> result = producer.send(producerRecord);
-			result.get();
-			logger.info("Sent record to kafka: {}", outboxRecord);
-			outboxRecord.setProcessed(now());
-			repository.update(outboxRecord);
+	private void processOutbox() {
+		repository.getUnprocessedRecords(BATCH_SIZE)
+				.stream()
+				.map(outboxRecord -> {
+					ProducerRecord<String, byte[]> producerRecord = toProducerRecord(outboxRecord);
+					return Tuple.of(outboxRecord, producer.send(producerRecord));
+				})
+				.toList() // collect to List before awaiting futures, to use producer internal batching
+				.forEach(tuple -> {
+					await(tuple.b);
+					OutboxRecord outboxRecord = tuple.a;
+					logger.info("Sent record to kafka: {}", outboxRecord);
+					outboxRecord.setProcessed(now());
+					repository.update(outboxRecord);
+				});
+	}
+
+	private static void await(Future<?> future) {
+		try {
+			future.get();
+		} catch (InterruptedException e) {
+			Thread.currentThread().interrupt();
+			throw new RuntimeException(e);
```

Comment:
Hmm... The `InterruptedException` would be thrown, if the current thread would get interrupted. How could that happen?
The `@PreDestroy` method `close()` is currently implemented with `executor.shutdown();`, which is documented with
> Initiates an orderly shutdown in which previously submitted tasks are executed
in contrast to `shutdownNow()`, which would interrupt running tasks.

I.e. in real life this should never happen. And if the `close()` method would be changed to use `shutdownNow()`, the service / process would already be shutting down. Triggering that again would not make sense to me.

Do you see another way, how the thread could be interrupted so that an `InterruptedException` is thrown?

Apart from that, one could question if the interrupted status of the thread is properly handled further up the call chain, but I'd argue that this could be a different issue. With the handling of the `InterruptedException` I simply followed the standard/default handling of it, that's why it's implemented as it is...

Reply:
Maybe I did not think sufficiently about this, but I thought as we could gracefully stop processing, but you are right it is simply not necessary and I am totally fine with this solution after a second thought. Thanks for your in depth explanation!